src/target/riscv/riscv.c

```diff
@@ -5941,50 +5942,32 @@ static int riscv_resume_go_all_harts(struct target *target)
 	return ERROR_OK;
 }
 
-int riscv_interrupts_disable(struct target *target, uint64_t irq_mask, uint64_t *old_mstatus)
+static int riscv_interrupts_disable(struct target *target, riscv_reg_t *old_mstatus)
 {
 	LOG_TARGET_DEBUG(target, "Disabling interrupts.");
-	struct reg *reg_mstatus = register_get_by_name(target->reg_cache,
-			"mstatus", true);
-	if (!reg_mstatus) {
-		LOG_TARGET_ERROR(target, "Couldn't find mstatus!");
-		return ERROR_FAIL;
+	riscv_reg_t current_mstatus;
+	int ret = riscv_reg_get(target, &current_mstatus, GDB_REGNO_MSTATUS);
+	if (ret != ERROR_OK) {
+		LOG_TARGET_ERROR(target, "Failed to read mstatus!");
+		return ret;
 	}
-
-	int retval = reg_mstatus->type->get(reg_mstatus);
-	if (retval != ERROR_OK)
-		return retval;
-
-	RISCV_INFO(info);
-	uint8_t mstatus_bytes[8] = { 0 };
-	uint64_t current_mstatus = buf_get_u64(reg_mstatus->value, 0, reg_mstatus->size);
-	buf_set_u64(mstatus_bytes, 0, info->xlen, set_field(current_mstatus,
-				irq_mask, 0));
-
-	retval = reg_mstatus->type->set(reg_mstatus, mstatus_bytes);
-	if (retval != ERROR_OK)
-		return retval;
-
 	if (old_mstatus)
 		*old_mstatus = current_mstatus;
-
-	return ERROR_OK;
+	return riscv_reg_set(target, GDB_REGNO_MSTATUS, current_mstatus & ~mstatus_ie_mask);
 }
 
-int riscv_interrupts_restore(struct target *target, uint64_t old_mstatus)
+static int riscv_interrupts_restore(struct target *target, riscv_reg_t old_mstatus)
 {
 	LOG_TARGET_DEBUG(target, "Restoring interrupts.");
-	struct reg *reg_mstatus = register_get_by_name(target->reg_cache,
-			"mstatus", true);
-	if (!reg_mstatus) {
-		LOG_TARGET_ERROR(target, "Couldn't find mstatus!");
-		return ERROR_FAIL;
+	riscv_reg_t current_mstatus;
+	int ret = riscv_reg_get(target, &current_mstatus, GDB_REGNO_MSTATUS);
+	if (ret != ERROR_OK) {
+		LOG_TARGET_ERROR(target, "Failed to read mstatus!");
+		return ret;
 	}
-
-	RISCV_INFO(info);
-	uint8_t mstatus_bytes[8];
-	buf_set_u64(mstatus_bytes, 0, info->xlen, old_mstatus);
-	return reg_mstatus->type->set(reg_mstatus, mstatus_bytes);
+	if ((current_mstatus & mstatus_ie_mask) != 0)
+		LOG_TARGET_WARNING(target, "Interrupt enable bits in mstatus changed!");
```

Comment:
Would you like this extended warning + hint? 

It should hopefully make it more clear to the user what has happened.

```suggestion
	if ((current_mstatus & mstatus_ie_mask) != 0) {
		LOG_TARGET_WARNING(target, "Interrupt enable bits in mstatus changed during single-step. OpenOCD might have affected the program flow by restoring the original interrupt enable bits.");
		LOG_TARGET_INFO(target, "Hint: Use 'riscv set_maskisr off' to prevent OpenOCD from touching mstatus during single-step.");
	}
```

Reply:
Done, only changed slightly to align with similar messages in https://github.com/riscv-collab/riscv-openocd/blob/riscv/src/target/riscv/riscv_semihosting.c#L130-L134